### PR TITLE
Make color palette extensible

### PR DESCRIPTION
**Pull request by kaelig**
This version takes advantage of the fact that if a colour doesn't exist as a use case or as a palette colour, it will return a "null":

```scss
$foo: null;
.el { color: $foo; }
```

Compiles to… nothing. Before it was compiling to "transparent" which is a destructive value.

- I've also made changes to the map syntax and warnings to make them more robust, compatible with all the edge cases and other 
- Tests: fixed tests to work with Sass 3.4

### CODE REVIEW

**Comment by AlbertoElias** [comment]
Why do you use quotes for colours now?

If you're going for a major release, feel free to finally remove all deprecated use cases.
Also, the travis test uses ruby-sass, but we're moving to node-sass

**Comment by kaelig** [comment]
Since a lot of dependants use o-colors it's a lot of work to deprecate stuff. Hopefully I'll have time later on to go through all components that use deprecated use cases and update them accordingly.

And quoting colours… because of this: https://twitter.com/kaelig/status/593828525863473152

**Comment by AlbertoElias** [comment]
@kaelig That's already been done. Use cases were deprecated AFTER adding them to the corresponding module

**Comment by kaelig** [comment]
Unfortunately a few modules still use the deprecated use cases in their latest versions… So it would mean we'd keep them locked into version 2 of o-colors. Not sure that's going to be a good decision considering it takes time to update the modules.

**Comment by AlbertoElias** [comment]
@kaelig Which ones? I'll work on updating them. And a major update should remove deprecated stuff, the deprecated use cases have been there for months, it's about time we removed them. And if we wait for v4, they're going to be there forever.

**Comment by kaelig** [comment]
Cool, sounds fair.

I propose this strategy while trying out v3.0.0-beta.x:
1. [ ] Patch v2 to make it work properly with newer versions of Sass, making it forward compatible with v3.0.0. Basically: the palette map will have both quoted keys (the new way) *and* non-quoted keys (the buggy way)
2. [ ] Patch all modules that use `oColorsSetUseCase` and `oColorsGetPaletteColor`: quote their parameters
3. [ ] Patch all calls to oColorsGetColorFor that use deprecated color use cases (a few really old modules still use them, such as o-author-alerts)
4. [ ] Release v3.0.0
5. [ ] Success!

**Comment by AlbertoElias** [comment]
Yes, I agree. Let me know when 1 is merged, and we can start doing 2 and 3.

**Comment by kaelig** [comment]
Another breaking change of v3:

```scss
$foo: oColorsGetPaletteColor(non-existing-color);
el {
  /* $foo: #{inspect($foo)} */
  background: mix(blue, $foo, 50%);
  color: $foo;
}
```

#### V2

Compiles to:

```css
el {
  /* $foo: transparent */
  background: rgba(0, 0, 255, 0.5);
  color: transparent;
}
```

#### V3
it just fails and throws this error message:

```
null is not a color for `mix'
```

----

I think it is a good thing because now we are warned when we use non existant colors, whereas before it failed silently. Any thoughts?

File: README.md [hunk 1]
```diff
@@ -37,8 +37,7 @@ The `oColorsGetColorFor` function takes three arguments:
 
 * **Use case list**: a list of colour use cases in order of preference.  The first one that is defined for the specified property will be returned
 * **Property**: The property that you want to use the colour for (background, border, or text).  Note that in contrast to the `oColorsFor` mixin, you must specify only one property.   Options are `background`, `border`, `text`, and `all`.
-* **Options**: A Sass *map* of additional options, all of which are optional, and may comprise:
-	* **default**: The name of a palette colour to return if none of the specified use cases are defined for the desired property.  May also be set to `null` or `undefined` to return that instead of the built in default (which is transparent)
+* **Default**: The name of a palette colour to return if none of the specified use cases are defined for the desired property.
```
Comment (triblondon) [inline README.md]: Why add this to the arg list?  The previous form was more extensible, wasn't it?

Comment (kaelig) [inline README.md]: Good point.

Actually after looking at it twice, I suggest that we remove that part in v3 since 0 components are using this parameter (only the demo does).

Comment (kaelig) [inline README.md]: Or maybe we can just make that parameter private because the mixin `oColorsFor` uses it

File: src/scss/_functions.scss [hunk 2]
```diff
@@ -10,31 +10,27 @@
 	@if (map-has-key($o-colors-palette, $name)) {
 		@return map-get($o-colors-palette, $name);
 	} @else {
-		@warn "Color name '" + $name + "' is not defined in the palette";
-		@return undefined;
+		@warn "Color name '#{inspect($name)}' is not defined in the palette";
+		@return null;
```
Comment (triblondon) [inline src/scss/_functions.scss]: Can you explain why this is needed?  Can't figure it out.

File: README.md (hunk 1 above)
Comment (triblondon) [inline README.md]: I would suggest leaving it as it was.  It's an extensible options map, so it doesn't get in the way of future things, is currently used internally, works, no need to change.

File: src/scss/_functions.scss [hunk 3]
```diff
@@ -44,33 +40,45 @@
...
+///
+/// // Assign a different fallback if the use case doesn't exist
+/// .foo { background-color: oColorsGetColorFor(foo, background, $default: pink-tint-1); }
 ///
 /// @param {list} $namelist
 /// @param {String} $property [all]
-/// @param {map} $options [()] - (default: xxx) name of a palette colour to return if none of the specified use cases are defined for the desired property
-@function oColorsGetColorFor($namelist, $property: all, $options: ()) {
+/// @param {Bool | String} $default [false] - name of a palette colour to return if none of the specified use cases are defined for the desired property
+@function oColorsGetColorFor($namelist, $property: all, $default: false) {
```
Comment (triblondon) [inline src/scss/_functions.scss]: Ditto prev comment here.

File: src/scss/_mixins.scss [hunk 4]
```diff
@@ -34,21 +34,12 @@
 /// @param {String|list} $propertyList [all]
 @mixin oColorsFor($useCaseList, $propertyList: all) {
 	@if ($propertyList == 'all' or index($propertyList, 'background')) {
-		$col: oColorsGetColorFor($useCaseList, background, (default:undefined));
-		@if ($col != undefined) {
-			background-color: $col;
-		}
+		background-color: oColorsGetColorFor($useCaseList, background, null);
```
Comment (triblondon) [inline src/scss/_mixins.scss]: Won't this output `background-color: null` if there's no color defined for the use case/prop combo?

File: src/scss/_functions.scss (hunk 2 above)
Comment (kaelig) [inline src/scss/_functions.scss]: Because Sass will transform some keywords (colors and other language keywords) if passed as is in an interpolated string.

e.g. if `$name` is `null`, it would output `Color name '' is not defined in the palette`.

File: README.md (hunk 1 above)
Comment (kaelig) [inline README.md]: Ok will revert it then.

File: src/scss/_mixins.scss (hunk 4 above)
Comment (kaelig) [inline src/scss/_mixins.scss]: It won't output anything at all (see PR description). This is a useful Sass thing: `prop: null` compiles to nothing.

**Comment by kaelig** [comment]
@triblondon I've restored the `$options` parameter in `oColorsGetColorFor`

**Comment by kaelig** [comment]
This is now working and all modules have had the time to update.

@samgiles @pornel Can you please let me know if there is a rise in build service errors?
